test/e2e/framework/pod/wait.go

```diff
@@ -99,17 +99,28 @@ func BeInPhase(phase v1.PodPhase) types.GomegaMatcher {
 	}).WithTemplate("Expected Pod {{.To}} be in {{format .Data}}\nGot instead:\n{{.FormattedActual}}").WithTemplateData(phase)
 }
 
-// WaitForPodsRunningReady waits up to timeout to ensure that all pods in
-// namespace ns are either running and ready, or failed but controlled by a
-// controller. Also, it ensures that at least minPods are running and
-// ready. It has separate behavior from other 'wait for' pods functions in
-// that it requests the list of pods on every iteration. This is useful, for
-// example, in cluster startup, because the number of pods increases while
-// waiting. All pods that are in SUCCESS state are not counted.
+// WaitForAlmostAllReady waits up to timeout for the following conditions:
+// 1. At least minPods Pods in Namespace ns are Running and Ready
+// 2. All Pods in Namespace ns are either Ready or Succeeded
+// 3. All Pods part of a ReplicaSet or ReplicationController in Namespace ns are Ready
+//
+// After the timeout has elapsed, an error is returned if and only if either of
+// the following conditions hold:
+//  1. The number of Pods in Namespace ns that are not Succeeded or Failed is less than minPods
+//  2. The number of Pods in Namespace ns that are not Ready, Succeeded, or Failed
+//     is greater than to allowedNotReadyPods
+//
+// Roughly speaking, this means that not "too many" Pods are not Ready, where the tolerance
+// for "too many" is controlled by allowedNotReadyPods.
+//
+// It is generally recommended to use WaitForPodsRunningReady instead of this function
+// whenever possible, because it is easier to understand. Similar to WaitForPodsRunningReady,
+// this function requests the list of pods on every iteration, making it useful for situations
+// where the set of Pods is likely changing, such as during cluster startup.
 //
 // If minPods or allowedNotReadyPods are -1, this method returns immediately
 // without waiting.
-func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns string, minPods, allowedNotReadyPods int32, timeout time.Duration) error {
+func WaitForAlmostAllPodsReady(ctx context.Context, c clientset.Interface, ns string, minPods, allowedNotReadyPods int, timeout time.Duration) error {
```

Comment:
This is almost identical to the previous `WaitForPodsRunningReady`. Most line changes are just cleanup/refactoring. The only actual change is the check at the end for whether to return an error, which I explain there.

This function is only called in the setup function in `e2e.go`, since that's the only place that it isn't clearly safe to change to cleaner function below.

Reply:
I think that is where the function originally came from, dating back to 2016 and earlier. That E2E tests then used it just "happened".

Reply:
I'm sure you're right! I looked through the history of this file for clues on how this function is actually supposed to work, but it's had the same basic structure since this file was created back in 2019 (clearly copy-and-pasted from some other file, but I stopped digging at that point).

Reply:
Just to be clear (because the threading is a bit odd): https://github.com/kubernetes/kubernetes/pull/124205#discussion_r1568311204 is about "let's keep WaitForAlmostAllPodsReady exactly as it was before".